### PR TITLE
Shared expenses

### PR DESCRIPTION
**Pull request by RezenkovD**

### CODE REVIEW

File: src/app/crud/crud_group.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        .filter(and_(UsersGroup.user_id == user_id, UsersGroup.group_id == group_id))
+        .one_or_none()
+    )
+
+
+def get_users_group_statistics(
+    db: Session,
+    group: GroupBase,
+    user_id: int,
+):
```
Comment (qqlexa) [inline src/app/crud/crud_group.py:89]: Return type annotation?
I see Union[list, bool] but it is not a good way to do this func.

Comment (qqlexa) [reply src/app/crud/crud_group.py:89]: Better is list or None, Optional[list]

File: src/app/crud/crud_group.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        db.refresh(db_users_group)
+
+    return s_group
+
+
+def get_group(db: Session, title: str):
+    return db.query(Group).filter_by(title=title).one_or_none()
+
+
+def check_user_group(db: Session, user_id: int, group_id: int):
```
Comment (qqlexa) [inline src/app/crud/crud_group.py:77]: Add return type annotation Optional[UsersGroup]

File: src/app/crud/crud_group.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    group: GroupBase,
+    user_id: int,
+):
+    s_group = get_group(db, group.title)
+    if s_group is not None:
+        user_group = check_user_group(db, user_id, s_group.id)
+        if user_group is not None:
+            db_query = db.query(UsersGroup).filter_by(group_id=s_group.id).all()
+            list_user_id = []
+            for x in range(len(db_query)):
```
[the remark this answers is not in the document]
Comment (RezenkovD) [reply src/app/crud/crud_group.py]: Thanks. This is the right decision.

File: src/app/crud/crud_group.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+
+def get_data_about_all_users_for_group(
+    db: Session,
+    group: GroupBase,
+    user_id: int,
+):
+    list_user_id = get_list_user_id_for_group(db, group, user_id)
+    if list_user_id:
+        all_users = []
```
[the remark this answers is not in the document]
Comment (RezenkovD) [reply src/app/crud/crud_group.py]: Great solution

File: src/app/crud/crud_group.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+def get_list_users_group(
+    db: Session,
+    user_id: int,
+    group_id: int,
+) -> list:
+    get_user_in_group = (
+        db.query(UserGroup)
+        .filter(and_(UserGroup.user_id == user_id, UserGroup.group_id == group_id))
+        .one_or_none()
```
Comment (qqlexa) [inline src/app/crud/crud_group.py:42]: Pythonic way is to catch it in try: and raise 
```python
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="You are not in this group!",
        )
```
in except

File: src/app/crud/crud_group.py (hunk 2 above)
Comment (RezenkovD) [reply src/app/crud/crud_group.py]: Yes

File: src/app/crud/crud_invitation.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    invitation = (
+        db.query(Invitation)
+        .filter(
+            and_(
+                Invitation.recipient_id == user_id,
+                Invitation.status == "awaiting",
+                Invitation.id == invitation_id,
+            )
+        )
+        .one_or_none()
```
Comment (qqlexa) [inline src/app/crud/crud_invitation.py:30]:
```python
try:
      invitation = ....one()
except: 
      raise HTTPException(
           status_code=starlette.status.HTTP_404_NOT_FOUND,
           detail="Invitation is not found",
      )
pass
```

File: src/app/models/invitation.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+from sqlalchemy.orm import relationship
+
+from app.db import Base
+
+
+class StatusDB(str, enum.Enum):
+    AWAITING = "awaiting"
+    ACCEPTED = "accepted"
+    DENIED = "denied"
+    OVERDUE = "overdue"
```
Comment (qqlexa) [inline src/app/models/invitation.py:14]: Is it necessary to use overdue in lowercase?

File: src/app/models/invitation.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    id = Column(Integer, primary_key=True, index=True)
+    status = Column(String, Enum(StatusDB), nullable=False)
+    sender_id = Column(Integer, ForeignKey("users.id"), nullable=False)
+    recipient_id = Column(Integer, ForeignKey("users.id"), nullable=False)
+    group_id = Column(Integer, ForeignKey("groups.id"), nullable=False)
+    time = Column(DateTime, default=datetime.datetime.utcnow)
+
+    group = relationship("Group", back_populates="invitations")
+    sender = relationship("User", foreign_keys=[sender_id])
+    recipient = relationship("User", foreign_keys=[recipient_id])
```
Comment (qqlexa) [inline src/app/models/invitation.py:34]: Is recipient list[User] or User?

File: src/app/crud/crud_invitation.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    if get_recipient_in_group:
+        raise HTTPException(
+            status_code=status.HTTP_405_METHOD_NOT_ALLOWED,
+            detail="The recipient is already in this group!",
+        )
+    get_invitation = (
+        db.query(Invitation)
+        .filter(
+            and_(
+                Invitation.status == "awaiting",
```
Comment (qqlexa) [inline src/app/crud/crud_invitation.py:107]: Why don't you use Enum.AWAITING?

File: src/app/crud/crud_invitation.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        db.query(UserGroup)
+        .filter(and_(UserGroup.user_id == user_id, UserGroup.group_id == group_id))
+        .one_or_none()
+    )
+    if get_user_in_group is None:
+        raise HTTPException(
+            status_code=status.HTTP_404_NOT_FOUND,
+            detail="You are not in this group!",
+        )
+    if get_user(db, recipient_id) is None:
```
Comment (qqlexa) [inline src/app/crud/crud_invitation.py:88]: It is better to save result to variable

File: src/app/models/invitation.py (hunk 8 above)
Comment (RezenkovD) [reply src/app/models/invitation.py:34]: User. As sender also User

File: src/app/crud/crud_invitation.py (hunk 9 above)
Comment (RezenkovD) [reply src/app/crud/crud_invitation.py:107]: it will be right

File: src/app/crud/crud_invitation.py (hunk 10 above)
Comment (RezenkovD) [reply src/app/crud/crud_invitation.py:88]: Okay

File: src/app/models/invitation.py (hunk 7 above)
Comment (RezenkovD) [reply src/app/models/invitation.py:14]: It will look better that way

**Pull request merged**